lfs/batcher.go

```diff
@@ -51,16 +55,22 @@ func (b *Batcher) acceptInput() {
 	exit := false
 
 	for {
-		batch := b.newBatch()
+		batch := make([]Transferable, 0, b.batchSize)
 	Loop:
 		for len(batch) < b.batchSize {
-			t, ok := <-b.input
-			if !ok {
-				exit = true // input channel was closed by Exit()
-				break Loop
+			select {
+			case t, ok := <-b.input:
+				if ok {
+					batch = append(batch, t)
+				} else {
+					exit = true // input channel was closed by Exit()
+					break Loop
+				}
+			case <-time.After(time.Millisecond * 300):
```

Comment:
I think an explicit flush adds more complexity but doesn't make things any more deterministic. The transfer loop blocks on `Next()`, so you'd presumably call `Flush()` when you `Add()` something (otherwise you're just pushing the timeout up). Transfers start as soon as the queue has received a batch worth. Say a transfer fails while the thing feeding the queue has added 52 items, so we `Add()` and `Flush()`, sending off a batch of 53 items. Then another fails and we send off a batch of 10 items.

I think the way to make it operate in a more deterministic manner is to withhold all failures until the end and then retry them.